### PR TITLE
feat(ssa): Multiple slice mergers

### LINKED ISSUE #2599: Multiple slice mergers requires equal size slices
### Aim

I have this program where the first if case is hit and the second else-case is hit:
```
fn merge_slices_mutate_two_ifs(x: Field, y: Field) -> [Field] {
    let mut slice = [0; 2];
    if x != y {
        slice = slice.push_back(y);
        slice = slice.push_back(x);
    } else {
        slice = slice.push_back(x);
    }
    if x == 20 {
        slice = slice.push_back(20);
    } else {
        slice = slice.push_back(15);
    }
    slice = slice.push_back(30);
    
    slice
}
```
I should be able to write it like this:
```
fn merge_slices_mutate_two_ifs(x: Field, y: Field) -> [Field] {
    let mut slice = [0; 2];
    if x != y {
        slice = slice.push_back(y);
        slice = slice.push_back(x);
    } else {
        slice = slice.push_back(x);
    }
    if x == 20 {
        slice = slice.push_back(20);
    } 
    slice = slice.push_back(15);
    slice = slice.push_back(30);
    
    slice
}
```

### Expected Behavior

We should have a resulting slice in both cases like this: `0x00, 0x00, 0x0a, 0x05, 0x0f, 0x1e`.

### Bug

However, when checking the values like such:
```
    let slice = merge_slices_mutate_two_ifs(x, y);
    assert(slice.len() == 6);
    assert(slice[3] == 5);
    assert(slice[4] == 15);
    assert(slice[5] == 30);
```
The second snippet has the wrong values assigned to the slice. Specifically, printing all fix values we get `0x00, 0x00, 0x0a, 0x05, 0x00, 0x0f`, and thus the assertion will fail. 

### To Reproduce

1. Copy the above snippets into the Noir program
2. Run the program and see it fail
3. If desired you can print each slice value instead of asserting: 
```
    dep::std::println(slice[0]);
    dep::std::println(slice[1]);    
    dep::std::println(slice[2]);
    dep::std::println(slice[3]);
    dep::std::println(slice[4]);
    dep::std::println(slice[5]);
```

### Installation Method

Compiled from source

### Nargo Version

nargo 0.11.0 (git version hash: 27288512da528c2f61c9418552c97a6d8642cafd, is dirty: true)

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves #2599

## Summary\*

This PR implements the following (this is also partially outlined in the issue discussion):

1. When we have a `push_back`, we need to check whether the length == capacity. If they are not equal we need to actually perform an array_set rather than making a new array. Simply adding an element to a slice which is the result of a slice merger would potentially result in the element being included after placeholder data in the slice's SSA value. When the slice is later accessed the placeholder data will potentially be accessed where the element that was pushed back is expected.
2. In order to conditionally array_set or make a new slice value we will need to expose the merging of values (the merge_values method in flattening) to the entire SSA crate rather than it being part of the flattening pass' context.
3. When we have a `pop_back` we no longer construct a new array by directly popping off of the the SSA array value that represents a slice. Instead, we simply array_get the flattened values of a slice and reduce the dynamically tracked length by one. For example in this program:
```rust
fn merge_slices_push_then_pop(x: Field, y: Field) -> [Field] {
    let mut slice = [0; 2];
    if x != y {
        slice = slice.push_back(y);
        slice = slice.push_back(x);
    } else {
        slice = slice.push_back(x);
    }
    slice = slice.push_back(30);
    if x == 20 {
        slice = slice.push_back(20);
    }
    let (slice, elem) = slice.pop_back();
    assert(elem == 30);
    let (slice, elem) = slice.pop_back();
    assert(elem == x);
    slice
}
```
When we get to the first pop_back we have a slice of (length 5, [0, 0, y, x, 30, 0]). If we just pop_back the slice value directly we will get 0. So instead we just do an array_get at len - 1 and leave the slice the exact same. The simplified result where we want to return (slice, elem) is just ((length 4, [0, 0, y, x, 30, 0]), 30). So the slice stays the same and we just reduce the length.

I posted under the issue description how I thought `insert` and `remove` would be the most challenging to implement, but it is in fact just `push_back` and `pop_back` as placeholder information is only every attached to the end of the slice. We can keep slices sound as long as we accurately track the slice length and its internal elements. This PR moves to further ignore any values that are part of a slice's SSA array value, but past the dynamically tracked slice length.

## Documentation

- [] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
I haven't finished going through this PR, but can you fix the conflicts?

**Comment by vezenovm** [comment]
> I haven't finished going through this PR, but can you fix the conflicts?

Yup will do, these look to be new ones that came up after the PR was made.

File: compiler/noirc_evaluator/src/ssa/ir/dfg.rs [hunk 1]
```diff
@@ -266,6 +266,32 @@ impl DataFlowGraph {
         self.intrinsics.get(&intrinsic)
     }
 
+    pub(crate) fn get_intrinsic_from_value(&self, value: ValueId) -> Option<Intrinsic> {
+        let intrinsics = self
+            .intrinsics
+            .iter()
+            .filter_map(
+                |(intrinsic, intrinsic_value_id)| {
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/ir/dfg.rs:274]: If you have a ValueId that refers to an intrinsic you should be able to retrieve it via:
```rs
match &dfg[value] {
    Value::Intrinsic(foo) => ....,
    _ => (),
}
```

File: compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs [hunk 2]
```diff
@@ -759,14 +538,34 @@ impl<'f> Context<'f> {
...
+        let block = self.inserter.function.entry_block();
+
+        let mut value_merger = ValueMerger::new(
+            &mut self.inserter.function.dfg,
+            block,
+            Some(&self.store_values),
+            Some(&self.outer_block_stores),
+        );
+
+        // Merging must occur in a separate loop as we cannot `*self` as mutable more than one at a time
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs:550]:
```suggestion
        // Merging must occur in a separate loop as we cannot borrow `self` as mutable while `value_merger` does
```

File: tooling/nargo_cli/tests/execution_success/slices/src/main.nr [hunk 3]
```diff
@@ -87,7 +87,7 @@ fn regression_2083() {
 // The parameters to this function must come from witness values (inputs to main)
 fn regression_merge_slices(x: Field, y: Field) {
     merge_slices_if(x, y);
-    merge_slices_else(x);
+    // merge_slices_else(x);
```
Comment (jfecher) [inline tooling/nargo_cli/tests/execution_success/slices/src/main.nr:90]: Is this breaking now?

Comment (vezenovm) [reply tooling/nargo_cli/tests/execution_success/slices/src/main.nr:90]: Nope I just forgot to uncomment while debugging, I have brought it back

File: compiler/noirc_evaluator/src/ssa/ir/dfg.rs (hunk 1 above)
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/ir/dfg.rs]: Ah yes, done

**Review by jfecher — APPROVED** [review]

**Pull request merged**